hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialProviderFactory.java

```diff
@@ -59,23 +61,40 @@ public abstract CredentialProvider createProvider(URI providerName,
     }
   }
 
+  /**
+   * Fail fast on any recursive load of credential providers, which can
+   * happen if the FS itself triggers the load.
+   * A simple boolean could be used here, as the synchronized block ensures
+   * that only one thread can be active at a time. An atomic is used
+   * for rigorousness.
+   */
+  private static final AtomicBoolean serviceLoaderLocked = new AtomicBoolean(false);
+
   public static List<CredentialProvider> getProviders(Configuration conf
                                                ) throws IOException {
-    List<CredentialProvider> result = new ArrayList<CredentialProvider>();
+    List<CredentialProvider> result = new ArrayList<>();
     for(String path: conf.getStringCollection(CREDENTIAL_PROVIDER_PATH)) {
       try {
         URI uri = new URI(path);
         boolean found = false;
         // Iterate serviceLoader in a synchronized block since
         // serviceLoader iterator is not thread-safe.
         synchronized (serviceLoader) {
-          for (CredentialProviderFactory factory : serviceLoader) {
-            CredentialProvider kp = factory.createProvider(uri, conf);
-            if (kp != null) {
-              result.add(kp);
-              found = true;
-              break;
+          try {
+            if (serviceLoaderLocked.getAndSet(true)) {
+              throw new PathIOException(path,
+                  "Recursive load of credential provider!");
```

Comment:
This could be a bit less implementation specific and provide more guidance - especially if this ends up being exposed to the client/CLI. I'll also nit-pick the use of bang there.

Reply:
will remove the !; not sure what else we could say.

One issue: is there any legit way this recursion could be triggered and not have it be a bug, e.g some cred provider wants to use credentials API itself?

Reply:
There was some reason that it wasn't appropriate to add this eliminateIncompatibleCredentialProviders call for all interactions. I don't recall what it was. I think some guidance like, "you may be trying to access a  credential within a filesystem that requires the same credentials" would go along way.